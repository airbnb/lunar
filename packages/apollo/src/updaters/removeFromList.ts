import { DocumentNode } from 'graphql';
import get from 'lodash/get';
import set from 'lodash/set';
import { DataProxy } from 'apollo-cache';
import prepareQuery from '../utils/prepareQuery';
import getQueryName from '../utils/getQueryName';

export default function removeFromList<Result, Vars = {}>(
  docOrQuery: DocumentNode | DataProxy.Query<Vars>,
  listPath: string,
  id: string | number,
  idName: string = 'id',
) {
  const query = prepareQuery<Vars>(docOrQuery);

  return (cache: DataProxy) => {
    const queryResult = cache.readQuery<Result>(query);
    const nextResult = { ...queryResult };
    const list = get(queryResult, listPath);

    if (typeof list === 'undefined' || !Array.isArray(list)) {
      if (__DEV__) {
        throw new TypeError(`"${getQueryName(query.query)}" list "${listPath}" is not an array.`);
      } else {
        return;
      }
    }

    set(
      nextResult,
      listPath,
      list.filter(item => item[idName] !== id),
    );

    cache.writeQuery({
      ...query,
      data: nextResult,
    });
  };
}
