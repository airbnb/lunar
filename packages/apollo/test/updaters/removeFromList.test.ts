import gql from 'graphql-tag';
import { InMemoryCache } from 'apollo-cache-inmemory';
import removeFromList from '../../src/updaters/removeFromList';

const QUERY = gql`
  query getSomething {
    something {
      id
      name
      things {
        id
        type
      }
    }
  }
`;

describe('removeFromList()', () => {
  let cache: InMemoryCache;

  beforeEach(() => {
    cache = new InMemoryCache();

    cache.writeQuery({
      query: QUERY,
      variables: { id: 123 },
      data: {
        something: {
          id: 123,
          name: 'Something',
          things: [
            { id: 1, type: 'A', __typename: 'thing' },
            { id: 2, type: 'B', __typename: 'thing' },
            { id: 3, type: 'C', __typename: 'thing' },
          ],
          __typename: 'something',
        },
      },
    });
  });

  it('errors if query does not exist in cache', () => {
    expect(() => {
      removeFromList(
        gql`
          query otherQuery {
            things {
              id
            }
          }
        `,
        'things',
        3,
      )(cache);
    }).toThrowErrorMatchingSnapshot();
  });

  it('errors if property name is not an array', () => {
    expect(() => {
      // Should be `something.things`
      removeFromList(QUERY, 'something.name', 3)(cache);
    }).toThrowErrorMatchingSnapshot();
  });

  it('removes item from list based on ID', () => {
    removeFromList(QUERY, 'something.things', 3)(cache);

    expect(cache.readQuery({ query: QUERY })).toEqual({
      something: {
        id: 123,
        name: 'Something',
        things: [
          { id: 1, type: 'A', __typename: 'thing' },
          { id: 2, type: 'B', __typename: 'thing' },
        ],
        __typename: 'something',
      },
    });
  });
});
