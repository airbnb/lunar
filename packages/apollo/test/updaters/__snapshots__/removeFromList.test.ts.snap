// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`removeFromList() errors if property name is not an array 1`] = `"\\"getSomething\\" list \\"something.name\\" is not an array."`;

exports[`removeFromList() errors if query does not exist in cache 1`] = `
"Can't find field things on object {
  \\"something\\": {
    \\"type\\": \\"id\\",
    \\"generated\\": false,
    \\"id\\": \\"something:123\\",
    \\"typename\\": \\"something\\"
  }
}."
`;
